### PR TITLE
Graceful shutdown

### PR DESCRIPTION
**Pull request by ignaciogoldchluk-yolo**
- Start `ConnectionServer` before `ConsumerSupervisor`
- Keep channels in `ConnectionServer` state, pass a reference to the consumers
- Close channels before closing connection when terminating `ConnectionServer`
- Add logs, Mermaid diagram, tests and documentation

### CODE REVIEW

File: lib/coney/connection_server.ex [hunk 1]
```diff
@@ -1,35 +1,47 @@
...
-  def start_link([consumers, [adapter: adapter, settings: settings, topology: topology]]) do
-    GenServer.start_link(__MODULE__, [consumers, adapter, settings, topology], name: __MODULE__)
+  def start_link([[adapter: adapter, settings: settings, topology: topology]]) do
+    GenServer.start_link(__MODULE__, [adapter, settings, topology], name: __MODULE__)
   end
 
-  def init([consumers, adapter, settings, topology]) do
+  @impl GenServer
+  def init([adapter, settings, topology]) do
     send(self(), :after_init)
```
Comment (yolo-vikram) [inline lib/coney/connection_server.ex]: Instead use: https://hexdocs.pm/elixir/1.12/GenServer.html#c:handle_continue/2

File: lib/coney/connection_server.ex [hunk 2]
```diff
@@ -85,29 +130,35 @@ defmodule Coney.ConnectionServer do
...
+    new_channels =
+      Enum.map(channels, fn {channel_ref, {consumer_pid, consumer, _dead_channel}} ->
+        new_channel = adapter.create_channel(conn)
+        adapter.subscribe(new_channel, consumer_pid, consumer)
 
-      {:ok, pid} = ConsumerSupervisor.start_consumer(consumer, subscribe_chan)
-      adapter.subscribe(subscribe_chan, pid, consumer)
+        {channel_ref, {consumer_pid, consumer, new_channel}}
+      end)
+      |> Map.new()
```
Comment (yolo-vikram) [inline lib/coney/connection_server.ex]:
```elixir
Enum.into(channels, %{}, fn {channel_ref, {consumer_pid, consumer, _dead_channel}} ->
      new_channel = adapter.create_channel(conn)
      adapter.subscribe(new_channel, consumer_pid, consumer)

      {channel_ref, {consumer_pid, consumer, new_channel}}
 end)
 ```

File: lib/coney/connection_server.ex [hunk 3]
```diff
@@ -40,28 +52,61 @@ defmodule Coney.ConnectionServer do
...
   def handle_info({:DOWN, _, :process, _pid, reason}, state) do
     ConnectionRegistry.disconnected(self())
     Logger.error("#{__MODULE__} (#{inspect(self())}) connection lost: #{inspect(reason)}")
-    rabbitmq_connect(state)
+    {:noreply, state |> rabbitmq_connect() |> update_channels()}
   end
 
+  @impl GenServer
   def terminate(_reason, %State{amqp_conn: conn, adapter: adapter} = _state) do
     :ok = adapter.close(conn)
```
Comment (yolo-vikram) [inline lib/coney/connection_server.ex:79]: we should close channels and the close conn?

File: lib/coney/connection_server.ex (hunk 2 above)
Comment (ignaciogoldchluk-yolo) [reply lib/coney/connection_server.ex]: We can do `Map.new(channels, fn ...)` too

File: test/support/fake_consumer.ex [hunk 4]
```diff
@@ -18,6 +18,7 @@ defmodule FakeConsumer do
       :reject -> :reject
       :reply -> {:reply, :data}
       :exception -> raise "Exception happen"
+      _other -> :ok
```
Comment (yolo-vikram) [inline test/support/fake_consumer.ex:21]: What this could be, can you please add inline comment?

File: lib/coney/connection_server.ex [hunk 5]
```diff
@@ -1,35 +1,51 @@
...
+  end
+
+  @impl true
+  def handle_continue(_continue_arg, state) do
+    {:noreply, rabbitmq_connect(state)}
   end
 
-  def confirm(channel, tag) do
-    GenServer.call(__MODULE__, {:confirm, channel, tag})
+  def confirm(channel_ref, tag) do
```
Comment (yolo-vikram) [inline lib/coney/connection_server.ex:44]: Add spec for these function

File: lib/coney/connection_server.ex [hunk 6]
```diff
@@ -40,28 +56,62 @@ defmodule Coney.ConnectionServer do
...
     ConnectionRegistry.disconnected(self())
     Logger.error("#{__MODULE__} (#{inspect(self())}) connection lost: #{inspect(reason)}")
-    rabbitmq_connect(state)
+    {:noreply, state |> rabbitmq_connect() |> update_channels()}
   end
 
-  def terminate(_reason, %State{amqp_conn: conn, adapter: adapter} = _state) do
+  @impl GenServer
+  def terminate(_reason, %State{amqp_conn: conn, adapter: adapter, channels: channels} = _state) do
+    close_channels(channels, adapter)
```
Comment (yolo-vikram) [inline lib/coney/connection_server.ex:78]: Maybe log here, Instead of every channel close?

**Review by yolo-vikram — APPROVED** [review]
Overall Looks good! 🚀 
Few minor comments.

File: lib/coney/connection_server.ex (hunk 6 above)
Comment (ignaciogoldchluk-yolo) [reply lib/coney/connection_server.ex:78]: Notifying on each channel close gives us more information on the number of workers, how long it takes to close each channel, etc.

File: test/support/fake_consumer.ex (hunk 4 above)
Comment (ignaciogoldchluk-yolo) [reply test/support/fake_consumer.ex:21]: It's just a passthrough for any message. `FakeConsumer` is used only for testing, and we generally do not care about the payload, we can test that with another consumer that expects a specific payload and sends a response accordingly.

File: lib/coney/connection_server.ex [hunk 7]
```diff
@@ -85,29 +140,43 @@ defmodule Coney.ConnectionServer do
...
          %State{
-           consumers: consumers,
            adapter: adapter,
            settings: settings,
            topology: topology
          } = state
        ) do
     conn = adapter.open(settings)
     adapter.init_topology(conn, topology)
-    start_consumers(consumers, adapter, conn)
```
Comment (edgarmiadzieles) [inline lib/coney/connection_server.ex:148]: Maybe we should move the Process.monitor from the adapter to here, seems like a more fitting context

**Review by edgarmiadzieles — APPROVED** [review]

File: lib/coney/connection_server.ex (hunk 6 above)
Comment (ignaciogoldchluk-yolo) [reply lib/coney/connection_server.ex:78]: Well, the application can have hundreds of channels open and we don't want to flood the logs. Modified to log a single message with the number of channels closed

**Pull request merged**